### ISSUE TITLE
loss becomes nan after 1st epoch while training 5 folds.

### ISSUE THREAD

**Issue by apthagowda97** [tensorflow/tpu#766]
According to this article [here](https://www.kaggle.com/c/jigsaw-multilingual-toxic-comment-classification/discussion/146242) its is suggested to use 
` tf.tpu.experimental.initialize_tpu_system(tpu)` before each fold to reinitialize the TPU but my train and val loss becomes nan from 2nd epoch of the 2nd fold. If I don't use that line I will get OOM exception.

```
def getmodel():
    with strategy.scope():
        model = tf.keras.Sequential([
            efn.EfficientNetB1(input_shape=(SIZE[0], SIZE[1], 3),weights='imagenet',pooling='avg',include_top=False),
            Dense(4, activation='softmax')])
        model.compile(
            optimizer='adam',loss = 'categorical_crossentropy',
            metrics=['categorical_accuracy',])
    return model

for i_fold, (train_idx, valid_idx) in enumerate(folds.split(train_df,train_y)):
    print("Fold {}/{}".format(i_fold + 1, N_FOLDS))
    
    tf.tpu.experimental.initialize_tpu_system(tpu)
    
    train = train_df.iloc[train_idx]
    train.reset_index(drop=True, inplace=True) 
    valid = train_df.iloc[valid_idx]
    valid.reset_index(drop=True, inplace=True)
    
    train_dataset = (tf.data.Dataset
    .from_tensor_slices(preprocess(train))
    .map(decode_image, num_parallel_calls=AUTO)
    .map(data_augment, num_parallel_calls=AUTO)
    .shuffle(SEED)
    .batch(BATCH_SIZE)
    .repeat()
    .prefetch(AUTO))

    valid_dataset = (tf.data.Dataset
    .from_tensor_slices(preprocess(valid))
    .map(decode_image, num_parallel_calls=AUTO)
    .batch(BATCH_SIZE)
    .cache()
    .prefetch(AUTO))
    
    test_dataset = (tf.data.Dataset
    .from_tensor_slices(preprocess(test_df,test=True))
    .map(decode_image, num_parallel_calls=AUTO)
    .batch(BATCH_SIZE))
    
    
    model = getmodel()
    steps = train.shape[0] // BATCH_SIZE
    history = model.fit(
        train_dataset, 
        epochs=EPOCHS, 
        callbacks=[lr_schedule],
        steps_per_epoch=steps,
        validation_data=valid_dataset)
    test_pred += model.predict(test_dataset)
```

Output:
```
Fold 2/5
Train for 22 steps, validate for 6 steps

Epoch 00001: LearningRateScheduler reducing learning rate to 0.0005.
Epoch 1/5
22/22 [==============================] - 154s 22s/step - loss: 0.9782 - categorical_accuracy: 0.6172 - val_loss: 0.9972 - val_categorical_accuracy: 0.5982

Epoch 00002: LearningRateScheduler reducing learning rate to 0.0004522542485937369.
Epoch 2/5
22/22 [==============================] - 29s 4s/step - loss: nan - categorical_accuracy: 0.2916 - val_loss: nan - val_categorical_accuracy: 0.2832
```

**Comment by apthagowda97** [comment]
Its a bug actually, my dataset contains 1821 images and when I used StratifiedKFold with folds = 5, the 1st fold has 1456 train and 365 valid images but rest of the fold(2nd to 5th) contains 1457 train and 364 valid images.

 `BATCH_SIZE = 8 * strategy.num_replicas_in_sync` (8*8 = 64)
`steps = train.shape[0] // BATCH_SIZE` (22)

1st fold ran successfully but loss became nan at the 2nd epoch of the 2nd fold. The problem is 1457 train images because it gives 22 steps which leave 49 images for the last batch but there were 8 TPU cores so 8 images at a time which leave 1 image at the last. 

I don't know why but because of this last single image my model loss became nan.

**Comment by gagika** [comment]
Can you try dropping the last partial batch?
batch(BATCH_SIZE, drop_remainder=True)
Or use batch(BATCH_SIZE).repeat() to avoid that last partial batch.

**Comment by apthagowda97** [comment]
Both of these methods works, and because of partial batch i was getting this error.
